src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs

```diff
@@ -77,36 +184,49 @@ public string PropertyPath()
         {
             StringBuilder sb = new StringBuilder("$");
 
-            for (int i = 0; i < _index; i++)
+            // If a continuation, always report back full stack.
+            int count = Math.Max(_count, _continuationCount);
+
+            for (int i = 0; i < count - 1; i++)
             {
                 AppendStackFrame(sb, _previous[i]);
             }
 
-            AppendStackFrame(sb, Current);
-            return sb.ToString();
-        }
+            if (_continuationCount == 0)
+            {
+                AppendStackFrame(sb, Current);
+            }
 
-        private void AppendStackFrame(StringBuilder sb, in WriteStackFrame frame)
-        {
-            // Append the property name.
-            string? propertyName = frame.JsonPropertyInfo?.PropertyInfo?.Name;
-            AppendPropertyName(sb, propertyName);
-        }
+            return sb.ToString();
 
-        private void AppendPropertyName(StringBuilder sb, string? propertyName)
-        {
-            if (propertyName != null)
+            void AppendStackFrame(StringBuilder sb, in WriteStackFrame frame)
             {
-                if (propertyName.IndexOfAny(ReadStack.SpecialCharacters) != -1)
+                // Append the property name.
+                string? propertyName = frame.DeclaredJsonPropertyInfo?.PropertyInfo?.Name;
+                if (propertyName == null)
                 {
-                    sb.Append(@"['");
-                    sb.Append(propertyName);
-                    sb.Append(@"']");
+                    // Attempt to get the JSON property name from the property name specified in re-entry.
+                    propertyName = frame.JsonPropertyNameAsString;
```

Comment:
Do we append enumerable's index? There is logic for that on `ReadStack` but not here.

Reply:
This PR didn't change that. JsonPath needs the count to match to JsonPath spec; there is no spec for ObjectPath and we don't know what index is being processed since we can't easily use .Count (or count the items) because that wouldn't return the index actually being processed. 

We'd have to keep a side integer and increment that for every element while writing -- adds cost and no one has asked for it.